### PR TITLE
Add option to serve WebAssembly app with multithreading

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
For detailed information, see the API review issue at https://github.com/dotnet/aspnetcore/issues/54071

### PR review notes

The product logic change is very small - it's just the changes in these two files:

 * https://github.com/dotnet/aspnetcore/pull/54062/files#diff-d41c3c2016bd6afc367d38c1969f9f1aeb4951227870eacaac43e1ac8f4aabec
 * https://github.com/dotnet/aspnetcore/pull/54062/files#diff-c3c483135ae3ca0235fe94c0e7beb3890b6c947e71a5a01e42be67da67a27339

Everything else is tests.

When merged, we can also mark https://github.com/dotnet/aspnetcore/issues/53723 as `test-fixed`.

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
# API review template

## Background and Motivation



## Proposed API

```diff
namespace Microsoft.AspNetCore.Components.WebAssembly.Server;

public sealed class WebAssemblyComponentsEndpointOptions
{
+    /// <summary>
+    /// Gets or sets a flag to determine whether to enable WebAssembly multithreading. If true,
+    /// the server will add headers similar to <code>Cross-Origin-Embedder-Policy: require-corp</code> and
+    /// <code>Cross-Origin-Opener-Policy: same-origin</code> on the response for the host page, because
+    /// this is required to enable the SharedArrayBuffer feature in the browser.
+    ///
+    /// Note that enabling this feature can restrict your ability to use other JavaScript APIs. For more
+    /// information, see <see href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements" />.
+    /// </summary>
+    public bool ServeMultithreadingHeaders { get; set; }
}
```

## Usage Examples

```cs
app.MapRazorComponents<App>()
    .AddInteractiveWebAssemblyRenderMode(options => { options.ServeMultithreadingHeaders = true; })
    .AddAdditionalAssemblies(typeof(MyWebAssemblyApp.Pages.Index).Assembly);
```

## Alternative Designs

Overall, since this feature is nascent and remains experimental, I'm erring on the side of making it more explicit and literal so people who turn it on know what they are getting into. In the future if we get to the point that wasm multithreading is so good that most people should enable it, we retain the option to streamline the usage.

### Alternative: Infer it automatically

We could try to infer this based on `<WasmEnableThreads>` in the wasm csproj, but:

1. There could be multiple wasm apps. What if they want different configurations?
2. Since this is an early and quite experimental feature, it's really good to force developers to opt in to serving the COOP headers explicitly. Doing so can restrict the availability of other JS APIs (e.g., can interfere with popups that might be used for auth flows) as described at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#security_requirements. It might turn out to be fine, but in case this causes issues for people, it's really good that the developer has a chance to discover what they are enabling and the possible side-effects of that.

### Alternative: Extension method

An alternative API would be making it an extension method on `RazorComponentsEndpointConventionBuilder`, e.g.:

    app.MapRazorComponents<App>().AddInteractiveWebAssemblyRenderMode().ServeWasmMultithreadingHeaders();

... but that would be very strange and ambiguous since you could also write:

    app.MapRazorComponents<App>().AddInteractiveServerRenderMode().ServeWasmMultithreadingHeaders();

I prefer it being an option on `AddInteractiveWebAssemblyRenderMode` since it is more specific and accurate, as it only applies within that case.

### Alternative: Naming

It would be more obvious to call it `EnableMultithreading` instead of `ServeMultithreadingHeaders`, but that would not so well serve the purpose of helping developer understand what exactly they are enabling and how it might have other side effects.

## Risks

Nothing obvious.

File: src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentsEndpointConventionBuilderExtensions.cs [hunk 1]
```diff
@@ -24,6 +26,24 @@ public static class WebAssemblyRazorComponentsEndpointConventionBuilderExtension
...
+            builder.Add(endpointBuilder =>
+            {
+                var needsCoopHeaders = endpointBuilder.Metadata.OfType<ComponentTypeMetadata>().Any() // e.g., /somecomponent
+                    || endpointBuilder.Metadata.OfType<WebAssemblyRenderModeWithOptions>().Any();     // e.g., /_framework/*
+                if (needsCoopHeaders && endpointBuilder.RequestDelegate is { } originalDelegate)
+                {
+                    endpointBuilder.RequestDelegate = httpContext =>
+                    {
+                        httpContext.Response.Headers["Cross-Origin-Embedder-Policy"] = "require-corp";
+                        httpContext.Response.Headers["Cross-Origin-Opener-Policy"] = "same-origin";
```
Comment (javiercn) [inline src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentsEndpointConventionBuilderExtensions.cs:40]: Can this be `same-origin-allow-popups`? If we are not having a way to configure this, we should at least be as lax as possible while enabling the feature.

Comment (SteveSandersonMS) [reply src/Components/WebAssembly/Server/src/Builder/WebAssemblyRazorComponentsEndpointConventionBuilderExtensions.cs:40]: No, unfortunately that's not enough for `SharedArrayBuffer` to be enabled. It has to be `same-origin`.

File: src/Components/test/E2ETest/Tests/ThreadingAppHostedTest.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure;
+using Microsoft.AspNetCore.Components.E2ETest.Infrastructure.ServerFixtures;
+using Microsoft.AspNetCore.E2ETesting;
+using Microsoft.AspNetCore.InternalTesting;
+using OpenQA.Selenium;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETest.Tests;
+
+public class ThreadingHostedAppTest
```
Comment (javiercn) [inline src/Components/test/E2ETest/Tests/ThreadingAppHostedTest.cs:13]: Is it worth having a dev server test?

Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Tests/ThreadingAppHostedTest.cs:13]: There already is one - see `ThreadingAppTest`.

**Review by javiercn — APPROVED** [review]

**Pull request merged**